File: Python Enrolment System - Basic FIT5136/src/Administrator.py
"""
Administrator - Represents an administrator who can manage students and units within the Student Enrollment System.

This class provides functionalities for administrators to add or remove students, manage unit enrollments for students,
and add new units to the system. It assumes interaction with a {StudentEnrollmentSystem} instance.

Author: Antoni Erdeg - antoni.erdeg@monash.edu
Version: 1.0
"""

class Administrator:
    def __init__(self, name):
        """
        Initializes the Administrator with a name.

        Args:
            name (str): The name of the administrator.
        """
        self.name = name

    def add_student(self, ses, student):
        """
        Adds a student to the Student Enrollment System.

        Args:
            ses (StudentEnrollmentSystem): The student enrollment system instance.
            student (Student): The student object to be added.

        Prints:
            str: Success message after adding the student.
        """
        ses.add_student(student)
        print("Student added successfully.")

    def add_unit_to_student(self, ses, student_id, unit):
        """
        Adds a unit to a student's list of units.

        Args:
            ses (StudentEnrollmentSystem): The student enrollment system instance.
            student_id (str): The ID of the student to whom the unit will be added.
            unit (Unit): The unit to be added to the student.

        This method calls the `add_unit_to_student` method of the `StudentEnrollmentSystem` class.
        """
        ses.add_unit_to_student(student_id, unit)

    def remove_unit_from_student(self, ses, student_id, unit_name):
        """
        Removes a unit from a student's list of units.

        Args:
            ses (StudentEnrollmentSystem): The student enrollment system instance.
            student_id (str): The ID of the student from whom the unit will be removed.
            unit_name (str): The name of the unit to be removed.

        Prints:
            str: Success or error message based on the operation outcome.
        """
        if ses.student_exists(student_id):
            # If the student exists in the system
            student = ses.students[student_id]
            units = student.list_units()
            if units is not None and unit_name in units:
                # If the unit is in the student's list of units
                student.remove_unit(unit_name)
                print(f"Unit '{unit_name}' removed successfully.")
            else:
                # If the unit is not found in the student's list
                print(f"Student isn't currently enrolled within Unit '{unit_name}'.")
        else:
            # If the student does not exist
            print("Student not found.")

    def add_unit(self, ses, unit):
        """
        Adds a unit to the Student Enrollment System.

        Args:
            ses (StudentEnrollmentSystem): The student enrollment system instance.
            unit (Unit): The unit object to be added.

        Prints:
            str: Success message after adding the unit.
        """
        ses.add_unit(unit)
        print("Unit added to the system successfully.")

File: Java Enrolment System - Basic FIT5136/src/Administrator.java
/**
 * Administrator - Represents an administrator who can manage students and units within the Student Enrollment System.
 * <p>
 * This class provides functionalities for administrators to add or remove students, manage unit enrollments for students,
 * and add new units to the system. It assumes interaction with a {@link StudentEnrollmentSystem} instance.
 * </p>
 *
 * <p>
 * @author Antoni Erdeg - antoni.erdeg@monash.edu<br>
 * @version: 1.0
 * </p>
 */
public class Administrator {
    // All final variables are constants and should be upper case.
    // Refer to PartTimeStudent for non-constant variable example i.e. courseAdvisor
    private final String NAME;

    /**
     * Constructs an Administrator with the specified name.
     *
     * @param name The name of the administrator.
     */
    public Administrator(String name) {
        this.NAME = name;
    }

    /**
     * Adds a student to the Student Enrollment System.
     *
     * @param ses The Student Enrollment System instance.
     * @param student The student to be added.
     */
    public void addStudent(StudentEnrollmentSystem ses, Student student) {
        ses.addStudent(student);
        System.out.println("Student added successfully.");
    }

    /**
     * Adds a unit to a specific student's list of units.
     *
     * @param ses The Student Enrollment System instance.
     * @param studentId The ID of the student.
     * @param unit The unit to be added.
     */
    public void addUnitToStudent(StudentEnrollmentSystem ses, String studentId, Unit unit) {
        // Check if student exists and then add the unit to the student's list
        if (ses.studentExists(studentId)) {
            ses.addUnitToStudent(studentId, unit);
            System.out.println("Unit added to student successfully.");
        } else {
            System.out.println("Student not found.");
        }
    }

    /**
     * Removes a unit from a specific student's list of units.
     *
     * @param ses The Student Enrollment System instance.
     * @param studentId The ID of the student.
     * @param unitName The name of the unit to be removed.
     */
    public void removeUnitFromStudent(StudentEnrollmentSystem ses, String studentId, String unitName) {
        // Check if student exists
        if (ses.studentExists(studentId)) {
            Student student = ses.getStudents().get(studentId);
            // List current units and check if the unit is enrolled
            student.listUnits();
            boolean unitFound = student.getUnits().stream()
                    .anyMatch(unit -> unit.getCode().equalsIgnoreCase(unitName));
            if (unitFound) {
                student.removeUnit(unitName);
                System.out.println("Unit removed successfully.");
            } else {
                System.out.println("Student isn't currently enrolled in that unit.");
            }
        } else {
            System.out.println("Student not found.");
        }
    }

    /**
     * Adds a new unit to the Student Enrollment System.
     *
     * @param ses The Student Enrollment System instance.
     * @param unit The unit to be added.
     */
    public void addUnit(StudentEnrollmentSystem ses, Unit unit) {
        ses.addUnit(unit);
        System.out.println("Unit added to the system successfully.");
    }

    /**
     * Returns the name of the administrator.
     *
     * @return The name of the administrator.
     */
    public String getName() {
        return NAME;
    }
}
